File: app/src/main/java/com/jrvermeer/psalter/UI/Activities/MainActivity.java
//package com.jrvermeer.psalter.UI.Activities;
//
//import com.jrvermeer.psalter.BuildConfig;
//import com.jrvermeer.psalter.Core.Contracts.IPagerCallbacks;
//import com.jrvermeer.psalter.Core.Models.LogEvent;
//import com.jrvermeer.psalter.Core.Models.Psalter;
//import com.jrvermeer.psalter.Core.Models.SearchMode;
//import com.jrvermeer.psalter.Infrastructure.Logger;
//import com.jrvermeer.psalter.Infrastructure.MediaServiceBinder;
//import com.jrvermeer.psalter.Infrastructure.SimpleStorage;
//import com.jrvermeer.psalter.Infrastructure.TutorialHelper;
//import com.jrvermeer.psalter.R;
//import com.jrvermeer.psalter.Core.Contracts.IPsalterRepository;
//import com.jrvermeer.psalter.Infrastructure.MediaService;
//import com.jrvermeer.psalter.Infrastructure.PsalterDb;
//import com.jrvermeer.psalter.UI.Adaptors.PsalterPagerAdapter;
//import com.jrvermeer.psalter.UI.Adaptors.PsalterSearchAdapter;
//
//import android.app.Service;
//import android.content.ComponentName;
//import android.content.Intent;
//import android.content.ServiceConnection;
//import android.content.res.Resources;
//import android.graphics.Color;
//import android.net.Uri;
//import android.os.Build;
//import android.os.IBinder;
//import android.support.design.widget.FloatingActionButton;
//import android.support.v4.media.MediaMetadataCompat;
//import android.support.v4.media.session.MediaControllerCompat;
//import android.support.v4.media.session.PlaybackStateCompat;
//import android.support.v4.view.MenuItemCompat;
//import android.support.v4.view.ViewPager;
//import android.support.v7.app.AppCompatActivity;
//import android.os.Bundle;
//import android.support.v7.widget.SearchView;
//import android.support.v7.widget.Toolbar;
//import android.text.InputType;
//import android.view.Menu;
//import android.view.MenuItem;
//import android.view.View;
//import android.widget.Button;
//import android.widget.ListView;
//import android.widget.TableLayout;
//import android.widget.TextView;
//import android.widget.Toast;
//
//import butterknife.BindView;
//import butterknife.ButterKnife;
//import butterknife.OnClick;
//import butterknife.OnItemClick;
//import butterknife.OnLongClick;
//import butterknife.OnPageChange;
//
//public class MainActivity extends AppCompatActivity implements
//        SearchView.OnQueryTextListener,
//        IPagerCallbacks {
//
//    private SimpleStorage storage = new SimpleStorage();
//    private SearchMode searchMode = SearchMode.Psalter;
//    private IPsalterRepository psalterRepository;
//    private TutorialHelper tutorials;
//    private MediaServiceBinder mediaService;
//
//    @BindView(R.id.viewpager) ViewPager viewPager;
//    @BindView(R.id.fabToggleScore) FloatingActionButton fabToggleScore;
//    @BindView(R.id.fab) FloatingActionButton fab;
//    @BindView(R.id.lvSearchResults) ListView lvSearchResults;
//    @BindView(R.id.toolbar) Toolbar toolbar;
//    @BindView(R.id.tableButtons) TableLayout tableButtons;
//    @BindView(R.id.searchBtn_Lyrics) Button  searchBtn_Lyrics;
//    @BindView(R.id.searchBtn_Psalm) Button searchBtn_Psalm;
//    @BindView(R.id.searchBtn_Psalter) Button searchBtn_Psalter;
//    MenuItem searchMenuItem;
//    SearchView searchView;
//
//    @Override
//    protected void onCreate(Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//        psalterRepository = new PsalterDb(this);
//
//        // must initialize theme before calling setContentView()
//        if(storage.isNightMode()) setTheme(R.style.AppTheme_Dark);
//        else setTheme(R.style.AppTheme_Light);
//
//        setContentView(R.layout.activity_main);
//        ButterKnife.bind(this);
//        setSupportActionBar(toolbar);
//
//        setFabChecked(storage.isScoreShown(), fabToggleScore, storage.isNightMode());
//        //initialize viewpager
//        PsalterPagerAdapter pagerAdapter = new PsalterPagerAdapter(this, psalterRepository,
//                this, storage.isScoreShown(), storage.isNightMode());
//        viewPager.setAdapter(pagerAdapter);
//        viewPager.setCurrentItem(storage.getPageIndex());
//
//        lvSearchResults.setAdapter(new PsalterSearchAdapter(this, psalterRepository));
//
//        tutorials = new TutorialHelper(this);
//        tutorials.showShuffleTutorial(fab);
//        tutorials.showScoreTutorial(fabToggleScore);
//    }
//
//    @Override
//    protected void onResume(){
//        super.onResume();
//        // initialize media service
//        bindService(new Intent(MainActivity.this, MediaService.class), mConnection, Service.BIND_AUTO_CREATE);
//    }
//
//    @Override
//    protected void onPause() {
//        super.onPause();
//        unbindService(mConnection);
//    }
//
//    @Override
//    public void onBackPressed(){
//        if(lvSearchResults.isShown()) {  collapseSearchView(); }
//        else super.onBackPressed();
//    }
//
//    @Override
//    public boolean onCreateOptionsMenu(final Menu menu) {
//        getMenuInflater().inflate(R.menu.menu_options, menu);
//
//        menu.findItem(R.id.action_nightmode).setChecked(storage.isNightMode());
//
//        searchMenuItem = menu.findItem(R.id.action_search);
//        searchView = (SearchView)searchMenuItem.getActionView();
//        searchView.setOnQueryTextListener(this);
//
//        MenuItemCompat.setOnActionExpandListener(searchMenuItem, new MenuItemCompat.OnActionExpandListener() {
//            @Override
//            public boolean onMenuItemActionExpand(MenuItem item) {
//                showSearchButtons();
//                return true;
//            }
//            @Override
//            public boolean onMenuItemActionCollapse(MenuItem item) {
//                hideSearchResultsScreen();
//                hideSearchButtons();
//                return true;
//            }
//        });
//        searchPsalter();
//        return true;
//    }
//
//    @Override
//    public boolean onQueryTextSubmit(String query) {
//        if(searchMode == SearchMode.Lyrics){
//            performStringSearch(query);
//        }
//        else {
//            try {
//                int number = Integer.parseInt(query);
//                if(searchMode == SearchMode.Psalter){
//                    if(1 <= number && number <= 434){
//                        collapseSearchView();
//                        goToPsalter(number);
//                        Logger.searchEvent(searchMode, query, null);
//                    }
//                    else{
//                        Toast.makeText(this, "Pick a number between 1 and 434", Toast.LENGTH_SHORT).show();
//                        return false;
//                    }
//                }
//                else if (searchMode == SearchMode.Psalm){
//                    if(1 <= number && number <= 150){
//                        performPsalmSearch(number);
//                    }
//                    else{
//                        Toast.makeText(this, "Pick a number between 1 and 150", Toast.LENGTH_SHORT).show();
//                        return false;
//                    }
//                }
//                else return false;
//
//            } catch (NumberFormatException ex){
//                return false;
//            }
//        }
//        searchView.clearFocus();
//
//        return true;
//    }
//    @Override
//    public boolean onQueryTextChange(String newText) {
//        if(searchMode == SearchMode.Lyrics && newText.length() > 1){
//            performStringSearch(newText);
//            return true;
//        }
//        return false;
//    }
//    @OnClick(R.id.searchBtn_Psalm)
//    public void searchPsalm(){
//        searchView.setInputType(InputType.TYPE_CLASS_NUMBER);
//        searchView.setQueryHint("Enter Psalm (1 - 150)");
//        searchMode = SearchMode.Psalm;
//        if(Build.VERSION.SDK_INT < 23){
//            searchBtn_Psalter.setTextAppearance(this, R.style.Button);
//            searchBtn_Psalm.setTextAppearance(this, R.style.Button_Selected);
//            searchBtn_Lyrics.setTextAppearance(this, R.style.Button);
//        }
//        else{
//            searchBtn_Psalter.setTextAppearance(R.style.Button);
//            searchBtn_Psalm.setTextAppearance(R.style.Button_Selected);
//            searchBtn_Lyrics.setTextAppearance(R.style.Button);
//        }
//    }
//    @OnClick(R.id.searchBtn_Lyrics)
//    public void searchLyrics() {
//        searchView.setInputType(InputType.TYPE_CLASS_TEXT);
//        searchView.setQueryHint("Enter search query");
//        searchMode = SearchMode.Lyrics;
//        if(Build.VERSION.SDK_INT < 23){
//            searchBtn_Psalter.setTextAppearance(this, R.style.Button);
//            searchBtn_Psalm.setTextAppearance(this, R.style.Button);
//            searchBtn_Lyrics.setTextAppearance(this, R.style.Button_Selected);
//        }
//        else {
//            searchBtn_Psalter.setTextAppearance(R.style.Button);
//            searchBtn_Psalm.setTextAppearance(R.style.Button);
//            searchBtn_Lyrics.setTextAppearance(R.style.Button_Selected);
//        }
//    }
//    @OnClick(R.id.searchBtn_Psalter)
//    public void searchPsalter(){
//        searchView.setInputType(InputType.TYPE_CLASS_NUMBER);
//        searchView.setQueryHint("Enter Psalter number (1 - 434)");
//        searchMode = SearchMode.Psalter;
//        if(Build.VERSION.SDK_INT < 23){
//            searchBtn_Psalter.setTextAppearance(this, R.style.Button_Selected);
//            searchBtn_Psalm.setTextAppearance(this, R.style.Button);
//            searchBtn_Lyrics.setTextAppearance(this, R.style.Button);
//        }
//        else{
//            searchBtn_Psalter.setTextAppearance(R.style.Button_Selected);
//            searchBtn_Psalm.setTextAppearance(R.style.Button);
//            searchBtn_Lyrics.setTextAppearance(R.style.Button);
//        }
//    }
//
//    @Override
//    public boolean onOptionsItemSelected(MenuItem item) {
//        int id = item.getItemId();
//        if(id == R.id.action_nightmode){
//            boolean nightMode = storage.toggleNightMode();
//            item.setChecked(nightMode);
//            Logger.changeTheme(nightMode);
//
//            if(Build.VERSION.SDK_INT == 23){ // framework bug in api 23 calling recreate inside onOptionsItemSelected.
//                finish();
//                startActivity(getIntent());
//            }
//            else recreate();
//        }
//        else if(id == R.id.action_random){
//            if(mediaService.isShuffling() && mediaService.isPlaying()){
//                mediaService.skipToNext();
//            }
//            else {
//                Psalter psalter = psalterRepository.getRandom();
//                viewPager.setCurrentItem(psalter.getId(), true);
//            }
//            Logger.event(LogEvent.GoToRandom);
//        }
//        else if(id == R.id.action_shuffle) {
//            tutorials.showShuffleReminderTutorial(fab);
//            shuffle();
//        }
//        else if(id == R.id.action_rate){
//            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=com.jrvermeer.psalter")));
//        }
//        else if(id == R.id.action_sendfeedback){
//            Intent emailIntent = new Intent(Intent.ACTION_SENDTO, Uri.fromParts(
//                    "mailto","jrvermeer.dev@gmail.com", null));
//            emailIntent.putExtra(Intent.EXTRA_SUBJECT, "Psalter App");
//            String body = "\n\n\n";
//            body += "---------------------------\n";
//            body += "App version: " + BuildConfig.VERSION_NAME + " (" + BuildConfig.VERSION_CODE + ")\n";
//            body += "Android version: " + Build.VERSION.RELEASE;
//            emailIntent.putExtra(Intent.EXTRA_TEXT, body);
//            startActivity(Intent.createChooser(emailIntent, "Send email..."));
//        }
//        else return false;
//
//        return true;
//    }
//
//    private void performStringSearch(String query){
//        showSearchResultsScreen();
//        ((PsalterSearchAdapter)lvSearchResults.getAdapter()).queryPsalter(query);
//        lvSearchResults.setSelectionAfterHeaderView();
//    }
//    private void performPsalmSearch(int psalm){
//        showSearchResultsScreen();
//        ((PsalterSearchAdapter)lvSearchResults.getAdapter()).getAllFromPsalm(psalm);
//        lvSearchResults.setSelectionAfterHeaderView();
//    }
//
//    private void goToPsalter(int psalterNumber){
//        Psalter psalter = psalterRepository.getPsalter(psalterNumber);
//        viewPager.setCurrentItem(psalter.getId(), true); //viewpager goes by index
//    }
//    private void showSearchResultsScreen(){
//        lvSearchResults.setVisibility(View.VISIBLE);
//        viewPager.setVisibility(View.GONE);
//        fab.hide();
//    }
//    private void hideSearchResultsScreen(){
//        lvSearchResults.setVisibility(View.GONE);
//        viewPager.setVisibility(View.VISIBLE);
//        fab.show();
//    }
//    private void collapseSearchView(){
//        if(searchMenuItem != null && searchMenuItem.isActionViewExpanded()) searchMenuItem.collapseActionView();
//    }
//    private void showSearchButtons(){
//        tableButtons.setVisibility(View.VISIBLE);
//        fabToggleScore.hide();
//        fab.hide();
//    }
//    private void hideSearchButtons(){
//        tableButtons.setVisibility(View.GONE);
//        fabToggleScore.show();
//        fab.show();
//    }
//
//    @OnClick(R.id.fabToggleScore)
//    public void toggleScore(){
//        PsalterPagerAdapter adapter = (PsalterPagerAdapter)viewPager.getAdapter();
//        if(storage.toggleScore()) {
//            adapter.showScore();
//            setFabChecked(true, fabToggleScore, storage.isNightMode());
//        }
//        else {
//            adapter.hideScore();
//            setFabChecked(false, fabToggleScore, storage.isNightMode());
//        }
//
//        int currentPage = viewPager.getCurrentItem();
//        viewPager.setAdapter(adapter);
//        viewPager.setCurrentItem(currentPage);
//
//        Logger.changeScore(storage.isScoreShown());
//    }
//
//    @OnClick(R.id.fab)
//    public void fabClick() {
//        if(mediaService.isPlaying()) mediaService.stop();
//        else {
//            Psalter psalter = getSelectedPsalter();
//            mediaService.play(this, psalter, false);
//            Logger.playbackStarted(psalter.getTitle(), false);
//        }
//    }
//    @OnLongClick(R.id.fab)
//    public boolean shuffle(){
//        Psalter psalter = getSelectedPsalter();
//        mediaService.play(this, psalter, true);
//        Logger.playbackStarted(psalter.getTitle(), true);
//
//        tutorials.showShuffleRandomTutorial(toolbar.findViewById(R.id.action_random));
//        return true;
//    }
//
//    @OnItemClick(R.id.lvSearchResults)
//        public void onItemClick(View view) {
//        try {
//            TextView tvNumber = ((PsalterSearchAdapter.ViewHolder) view.getTag()).getTvNumber();
//            int num = Integer.parseInt(tvNumber.getText().toString());
//
//            //log event before collapsing searchview, so we can log the query text
//            Logger.searchEvent(searchMode, searchView.getQuery().toString(), String.valueOf(num));
//
//            collapseSearchView();
//            goToPsalter(num);
//        } catch (Exception ex) {
//            Toast.makeText(MainActivity.this, ex.getMessage(), Toast.LENGTH_SHORT).show();
//        }
//    }
//
//    @Override
//    public void pageCreated(View page, int position) {
//        if(position == viewPager.getCurrentItem()){
//            tutorials.showGoToPsalmTutorial(page.findViewById(R.id.tvPagerPsalm));
//        }
//    }
//
//    @OnPageChange(value = R.id.viewpager)
//    public void onPageSelected(int index) {
//        storage.setPageIndex(index);
//        if(mediaService != null && mediaService.isPlaying() && mediaService.getCurrentMediaId() != index){ // if we're playing audio of a different #, stop it
//            mediaService.stop();
//        }
//    }
//
//    private Psalter getSelectedPsalter(){
//        return psalterRepository.getIndex(viewPager.getCurrentItem());
//    }
//    ServiceConnection mConnection = new ServiceConnection() {
//        @Override
//        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
//            Logger.d("MediaService connected to Activity");
//            mediaService = (MediaServiceBinder)iBinder;
//            mediaService.registerCallback(callback);
//        }
//
//        @Override
//        public void onServiceDisconnected(ComponentName componentName) {
//            Logger.d("MediaService disconnected from Activity");
//        }
//    };
//
//    MediaControllerCompat.Callback callback = new MediaControllerCompat.Callback() {
//        @Override
//        public void onPlaybackStateChanged(PlaybackStateCompat state) {
//            Logger.d("Playback state changed: " + state.getState());
//            if(state.getState() == PlaybackStateCompat.STATE_PLAYING){
//                fab.setImageResource(R.drawable.ic_stop_white_24dp);
//            }
//            else fab.setImageResource(R.drawable.ic_play_arrow_white_24dp);
//
//            if(fab.isShown()) {  // stupid ass bug, setting images fails after toggling night mode. https://stackoverflow.com/a/52158081
//                fab.hide();
//                fab.show();
//            }
//        }
//
//        @Override
//        public void onMetadataChanged(MediaMetadataCompat metadata) {
//            Logger.d("Metadata changed");
//            if(mediaService.isPlaying()){
//                int id = Integer.parseInt(metadata.getDescription().getMediaId());
//                viewPager.setCurrentItem(id);
//            }
//        }
//    };
//
//    private void setFabChecked(boolean checked, FloatingActionButton fab, boolean nightMode){
//        int color;
//        Resources res = getResources();
//        if(checked && nightMode) color = Color.WHITE;
//        else if(checked) color = res.getColor(R.color.colorAccent);
//        else if(nightMode) color = res.getColor(R.color.colorUnselectedInverse);
//        else color = res.getColor(R.color.colorUnselected);
//        fab.getDrawable().mutate().setTint(color);
//    }
//}

File: app/src/main/java/com/jrvermeer/psalter/ui/MainActivity.kt
package com.jrvermeer.psalter.ui

import kotlinx.android.synthetic.main.activity_main.*

import com.jrvermeer.psalter.models.*
import com.jrvermeer.psalter.infrastructure.*
import com.jrvermeer.psalter.ui.adaptors.*

import android.app.Service
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.PlaybackStateCompat
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.viewpager.widget.ViewPager
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import android.text.InputType
import android.util.TypedValue
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatDelegate
import androidx.appcompat.widget.AppCompatButton
import com.google.android.material.snackbar.Snackbar
import com.jrvermeer.psalter.*
import com.jrvermeer.psalter.helpers.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import androidx.lifecycle.LifecycleOwner
import kotlinx.coroutines.cancel


class MainActivity : AppCompatActivity(), CoroutineScope by MainScope(), LifecycleOwner {

    private lateinit var storage: StorageHelper
    private lateinit var rateHelper: RateHelper
    private lateinit var instant: InstantHelper
    private lateinit var psalterDb: PsalterDb
    private lateinit var tutorials: TutorialHelper
    private lateinit var searchMenuItem: MenuItem
    private lateinit var searchView: SearchView
    private lateinit var downloader: DownloadHelper
    private var mediaService: MediaServiceBinder? = null
    private var menu: Menu? = null

    private val selectedPsalter get() = psalterDb.getIndex(viewPager.currentItem)

    override fun onCreate(savedInstanceState: Bundle?) {
        Logger.init(this)
        storage = StorageHelper(this)
        downloader = DownloadHelper(this, storage)
        rateHelper = RateHelper(this, storage)
        tutorials = TutorialHelper(this)
        instant = InstantHelper(this)
        psalterDb = PsalterDb(this, this, downloader)

        lifecycle.addObserver(psalterDb)

        // must be done before super(), or onCreate() will be called twice and tutorials won't work
        AppCompatDelegate.setDefaultNightMode(if(storage.nightMode) AppCompatDelegate.MODE_NIGHT_YES else AppCompatDelegate.MODE_NIGHT_NO)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initViews()

        storage.launchCount++
        // instant.transferInstantAppData() doesn't work anyways

        tutorials.showScoreTutorial(fabToggleScore)
        tutorials.showShuffleTutorial(fab)
    }

    override fun onStart() {
        super.onStart()
        // initialize media service
        bindService(Intent(this@MainActivity, MediaService::class.java), mConnection, Service.BIND_AUTO_CREATE)
    }

    override fun onStop() {
        mediaService?.unregisterCallbacks()
        unbindService(mConnection)
        super.onStop()
    }

    override fun onDestroy() {
        super.onDestroy()
        cancel()
    }
    override fun onBackPressed() {
        if (lvSearchResults.isShown) {
            collapseSearchView()
        } else
            super.onBackPressed()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        this.menu = menu
        menuInflater.inflate(R.menu.menu_options, menu)
        menu.findItem(R.id.action_nightMode).isChecked = storage.nightMode
        initSearchView(menu)
        searchMode = SearchMode.Psalter
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu?): Boolean {
        menu?.findItem(R.id.action_downloadAll)?.isVisible = !storage.allMediaDownloaded && !instant.isInstantApp
        if(storage.fabLongPressCount > 7) menu?.findItem(R.id.action_shuffle)?.isVisible = false
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_nightMode -> toggleNightMode()
            R.id.action_random -> goToRandom()
            R.id.action_shuffle -> shuffle(true)
            R.id.action_sendFeedback -> startActivity(IntentHelper.FeedbackIntent)
            R.id.action_downloadAll -> queueDownloads()
            R.id.action_favorites -> showFavorites()
            R.id.action_fontSize_small -> setFontScale(.875f)
            R.id.action_fontSize_normal -> setFontScale(1f)
            R.id.action_fontSize_big -> setFontScale(1.125f)
            R.id.action_fontSize_huge -> setFontScale(1.25f)
            else -> return false
        }
        return true
    }

    private fun setFontScale(scale: Float) {
        val adapter = viewPager.adapter as PsalterPagerAdapter
        storage.textScale = scale
        adapter.updateViews()
        updateViewPagerTitleTextSize()
    }

    private fun queueDownloads(){
        downloader.offlinePrompt(psalterDb, this) { msg -> snack(msg) }
    }

    private fun goToRandom() {
        if (mediaService?.isShuffling == true && mediaService?.isPlaying == true) {
            Logger.skipToNext(selectedPsalter)
            mediaService?.skipToNext()
        } else {
            Logger.event(LogEvent.GoToRandom)
            val next = psalterDb.getRandom()
            viewPager.setCurrentItem(next.id, true)
        }
        rateHelper.showRateDialogIfAppropriate()
    }

    private fun toggleNightMode() {
        storage.nightMode = !storage.nightMode
        Logger.changeTheme(storage.nightMode)
        recreateSafe()
    }

    private fun performPsalterSearch(psalterNumber: Int){
        if (psalterNumber in 1..434) {
            collapseSearchView()
            goToPsalter(psalterNumber)
            Logger.searchPsalter(psalterNumber)
        }
        else snack("Pick a number between 1 and 434")
    }
    private fun performLyricSearch(query: String, logEvent: Boolean) {
        showSearchResultsScreen()
        (lvSearchResults.adapter as PsalterSearchAdapter).queryPsalter(query)
        lvSearchResults.setSelectionAfterHeaderView()
        if(logEvent) Logger.searchLyrics(query)
    }
    private fun performPsalmSearch(psalm: Int) {
        if (psalm in 1..150) {
            showSearchResultsScreen()
            (lvSearchResults.adapter as PsalterSearchAdapter).getAllFromPsalm(psalm)
            lvSearchResults.setSelectionAfterHeaderView()
            Logger.searchPsalm(psalm)
        }
        else snack("Pick a number between 1 and 150")
    }

    private fun showFavorites(): Boolean {
        if(!psalterDb.getFavorites().any()) {
            tutorials.showAddToFavoritesTutorial(fabToggleFavorite)
            return false
        }

        searchMode = SearchMode.Favorites
        menu?.hideAll()
        showSearchResultsScreen()

        (lvSearchResults.adapter as PsalterSearchAdapter).showFavorites()
        lvSearchResults.setSelectionAfterHeaderView()
        return true
    }


    private fun showSearchButtons() {
        tableButtons.show()
        hideFabs()
    }
    private fun hideSearchButtons() {
        tableButtons.hide()
        showFabs()
    }

    private fun showFabs(){
        fabToggleScore.show()
        fabToggleFavorite.show()
        fab.show()
    }
    private fun  hideFabs(){
        fabToggleScore.hide()
        fabToggleFavorite.hide()
        fab.hide()
    }

    private fun showSearchResultsScreen() {
        lvSearchResults.show()
        viewPager.hide()
        hideFabs()
    }
    private fun hideSearchResultsScreen() {
        lvSearchResults.hide()
        viewPager.show()
        showFabs()
    }

    private fun collapseSearchView() {
        if (searchMenuItem.isActionViewExpanded) searchMenuItem.collapseActionView()
        hideSearchResultsScreen()
        invalidateOptionsMenu() // rebuild menu
    }

    private fun goToId(id: Int){
        viewPager.setCurrentItem(id, true) //viewpager goes by index
    }
    private fun goToPsalter(psalterNumber: Int) {
        val psalter = psalterDb.getPsalter(psalterNumber)!!
        goToId(psalter.id)
    }

    private fun toggleScore() {
        val adapter = viewPager.adapter as PsalterPagerAdapter
        storage.scoreShown = !storage.scoreShown
        adapter.updateViews()
        fabToggleScore.isSelected = storage.scoreShown
        Logger.changeScore(storage.scoreShown)
    }

    private fun toggleFavorite() {
        psalterDb.toggleFavorite(selectedPsalter!!)
        fabToggleFavorite.isSelected = !fabToggleFavorite.isSelected
        tutorials.showViewFavoritesTutorial(toolbar as Toolbar)
    }

    private fun togglePlay() {
        if (mediaService?.isPlaying == true) {
            mediaService?.stop()
            rateHelper.showRateDialogIfAppropriate()
        }
        else {
            val psalter = selectedPsalter
            mediaService?.play(psalter!!, false)
            mediaService?.startService(this)
        }
    }

    private fun shuffle(showLongPressTutorial: Boolean = false): Boolean {
        if(showLongPressTutorial) tutorials.showShuffleReminderTutorial(fab)
        else storage.fabLongPressCount++

        mediaService?.play(selectedPsalter!!, true)
        mediaService?.startService(this)
        rateHelper.showRateDialogIfAppropriate()
        return true
    }

    private fun onItemClick(view: View) {
        try {
            val tvNumber = (view.tag as PsalterSearchAdapter.ViewHolder).tvNumber
            val tvId = (view.tag as PsalterSearchAdapter.ViewHolder).tvId
            val num = Integer.parseInt(tvNumber!!.text.toString())
            val id = Integer.parseInt(tvId!!.text.toString())

            //log event before collapsing searchview, so we can log the query text
            Logger.searchEvent(searchMode, searchView.query.toString(), num)

            collapseSearchView()
            goToId(id)
            rateHelper.showRateDialogIfAppropriate()
        }
        catch (ex: Exception) {
            Logger.e("Error in MainActivity.onItemClick", ex)
        }
    }

    private fun onPageSelected(index: Int, view: View?) {
        Logger.d("Page selected: $index")
        storage.pageIndex = index
        fabToggleFavorite.isSelected = selectedPsalter?.isFavorite ?: false
        if (mediaService?.isPlaying == true && mediaService?.currentMediaId != index) { // if we're playing audio of a different #, stop it
            mediaService?.stop()
        }
    }

    private var mConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
            Logger.d("MediaService bound")
            mediaService = iBinder as MediaServiceBinder
            mediaService?.registerCallbacks(callback)
        }

        override fun onServiceDisconnected(componentName: ComponentName) {
            // this method IS NOT CALLED on unbindService(), only when service crashes or something. So it's pretty useless and misleading
        }
    }

    private var callback = object : MediaServiceCallbacks () {
        override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
            fab.isSelected = mediaService?.isPlaying ?: false
        }

        override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
            if (mediaService?.isPlaying == true) viewPager.currentItem = metadata!!.description.mediaId!!.toInt()
        }

        override fun onAudioUnavailable(psalter: Psalter) {
            snack("Audio unavailable for ${psalter.title}")
        }

        override fun onBeginShuffling() {
            snack("Shuffling", MessageLength.Short, "Skip") {
                mediaService?.skipToNext()
            }
        }
    }

    private lateinit var _searchMode: SearchMode
    private var searchMode
        get() = _searchMode
        set(mode) {
            _searchMode = mode
            searchView.setQuery("", false)
            when(mode) {
                SearchMode.Lyrics -> {
                    searchView.inputType = InputType.TYPE_CLASS_TEXT
                    searchView.queryHint = "Enter search query"
                    searchBtn_Lyrics.deselect(searchBtn_Psalter, searchBtn_Psalm)
                }
                SearchMode.Psalm -> {
                    searchView.inputType = InputType.TYPE_CLASS_NUMBER
                    searchView.queryHint = "Enter Psalm (1 - 150)"
                    searchBtn_Psalm.deselect(searchBtn_Psalter, searchBtn_Lyrics)
                }
                SearchMode.Psalter -> {
                    searchView.inputType = InputType.TYPE_CLASS_NUMBER
                    searchView.queryHint = "Enter Psalter number (1 - 434)"
                    searchBtn_Psalter.deselect(searchBtn_Psalm, searchBtn_Lyrics)
                }
            }
        }

    private fun initViews() {
        setSupportActionBar(toolbar as Toolbar)

        fab.setOnClickListener { togglePlay() }
        fab.setOnLongClickListener { shuffle() }

        fabToggleScore.setOnClickListener { toggleScore() }
        fabToggleScore.isSelected = storage.scoreShown

        fabToggleFavorite.setOnClickListener { toggleFavorite() }
        fabToggleFavorite.isSelected = psalterDb.getIndex(storage.pageIndex)?.isFavorite ?: false

        searchBtn_Psalm.setOnClickListener { searchMode = SearchMode.Psalm }
        searchBtn_Lyrics.setOnClickListener { searchMode = SearchMode.Lyrics }
        searchBtn_Psalter.setOnClickListener { searchMode = SearchMode.Psalter }

        val viewPagerAdapter = PsalterPagerAdapter(this, this, psalterDb, downloader, storage)
        viewPager.adapter = viewPagerAdapter
        viewPager.currentItem = storage.pageIndex
        updateViewPagerTitleTextSize()
        viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrollStateChanged(state: Int) {}
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
            override fun onPageSelected(position: Int) {
                this@MainActivity.onPageSelected(position, viewPagerAdapter.getView(position))
            }
        })

        lvSearchResults.adapter = PsalterSearchAdapter(this, psalterDb, storage)
        lvSearchResults.setOnItemClickListener { _, view, _, _ -> onItemClick(view) }
    }
    private fun updateViewPagerTitleTextSize(){
        viewPagerTitle.setTextSize(TypedValue.COMPLEX_UNIT_SP, 18 * storage.textScale)
    }
    private fun initSearchView(menu: Menu){
        searchMenuItem = menu.findItem(R.id.action_search)
        searchView = searchMenuItem.actionView as SearchView
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                when (searchMode) {
                    SearchMode.Lyrics -> performLyricSearch(query ?: "", true)
                    SearchMode.Psalm -> performPsalmSearch(query?.toIntOrNull() ?: 0)
                    SearchMode.Psalter -> performPsalterSearch(query?.toIntOrNull() ?: 0)
                }
                searchView.clearFocus()
                return true
            }

            override fun onQueryTextChange(query: String?): Boolean {
                if (searchMode == SearchMode.Lyrics && query != null && query.length > 1) {
                    performLyricSearch(query, false)
                    return true
                }
                return false
            }
        })

        searchMenuItem.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
            override fun onMenuItemActionExpand(item: MenuItem): Boolean {
                showSearchButtons()
                menu.hideAll(searchMenuItem)
                return true
            }

            override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                hideSearchButtons()
                invalidateOptionsMenu()
                return true
            }
        })
    }

    private fun snack(msg: String, len: MessageLength = MessageLength.Long){
        Snackbar.make(mainCoordLayout, msg, len.forSnack()).show()
    }
    private fun snack(msg: String, len: MessageLength, action: String, onClick: () -> Unit){
        Snackbar.make(mainCoordLayout, msg, len.forSnack())
                .setAction(action) { onClick() }.show()
    }

    private fun AppCompatButton.deselect(vararg deselect: AppCompatButton){
        this.isSelected = true
        deselect.forEach { btn -> btn.isSelected = false }
    }

    private fun Menu.hideAll(except: MenuItem? = null) {
        for (i in 0 until this.size()) {
            val item = this.getItem(i)
            if (item != except) item.isVisible = false
        }
    }
}
